src/connection.cpp

```diff
@@ -108,21 +105,33 @@ void Connection::accept(Protocol_ptr protocol)
 {
 	this->protocol = protocol;
 	g_dispatcher.addTask(createTask(std::bind(&Protocol::onConnect, protocol)));
+	connectionState = CONNECTION_STATE_CONNECTING_STAGE2;
 
 	accept();
 }
 
 void Connection::accept()
 {
+	if (connectionState == CONNECTION_STATE_PENDING) {
+		connectionState = CONNECTION_STATE_CONNECTING_STAGE1;
+	}
+
 	std::lock_guard<std::recursive_mutex> lockClass(connectionLock);
 	try {
 		readTimer.expires_from_now(std::chrono::seconds(CONNECTION_READ_TIMEOUT));
 		readTimer.async_wait(std::bind(&Connection::handleTimeout, std::weak_ptr<Connection>(shared_from_this()), std::placeholders::_1));
 
-		// Read size of the first packet
-		boost::asio::async_read(socket,
-		                        boost::asio::buffer(msg.getBuffer(), NetworkMessage::HEADER_LENGTH),
-		                        std::bind(&Connection::parseHeader, shared_from_this(), std::placeholders::_1));
+		if (!receivedLastChar && receivedName && connectionState == CONNECTION_STATE_CONNECTING_STAGE2) {
+			// Read size of the first packet
+			boost::asio::async_read(socket,
+				boost::asio::buffer(msg.getBuffer(), 1),
+				std::bind(&Connection::parseHeader, shared_from_this(), std::placeholders::_1));
+		} else {
+			// Read size of the first packet
+			boost::asio::async_read(socket,
+				boost::asio::buffer(msg.getBuffer(), NetworkMessage::HEADER_LENGTH),
+				std::bind(&Connection::parseHeader, shared_from_this(), std::placeholders::_1));
```

Comment:
You can use a lambda here, I think it will be way cleaner, e.g.

```cpp
		auto bufferLength = !receivedLastChar && receivedName && connectionState == CONNECTION_STATE_CONNECTING_STAGE2 ?  1 : NetworkMessage::HEADER_LENGTH
			// Read size of the first packet
		boost::asio::async_read(socket,
			boost::asio::buffer(msg.getBuffer(), bufferLength),
			[=, connection = shared_from_this()](auto buf) { connection->parseHeader(buf); });
```

Not 100% sure about the syntax but you'll get the idea

Reply:
I don't understand this enough to make that change. I'm not that good at c++

Reply:
I believe it is right now, just copy it over the whole if/else block